controller/api/views.py

```diff
@@ -412,19 +410,28 @@ def list(self, request, **kwargs):
                      if u.has_perm(perm_name, app)]
         return Response({'users': usernames})
 
-    @permission_classes([permissions.IsOwnerOrAdmin])
     def create(self, request, **kwargs):
         app = self.get_object()
+        if not permissions.IsOwnerOrAdmin.has_object_permission(permissions.IsOwnerOrAdmin(),
+                                                                request, self, app):
+            raise PermissionDenied()
+
         user = get_object_or_404(User, username=request.data['username'])
         assign_perm(self.perm, user, app)
         models.log_event(app, "User {} was granted access to {}".format(user, app))
         return Response(status=status.HTTP_201_CREATED)
 
-    @permission_classes([permissions.IsOwnerOrAdmin])
     def destroy(self, request, **kwargs):
-        app = self.get_object()
+        app = get_object_or_404(models.App, id=self.kwargs['id'])
```

Comment:
how come this call needs to change? Won't `get_object()` [do the right thing](https://github.com/tomchristie/django-rest-framework/blob/3e1687282ba39b0166b7c17e39cdd3a2be7e2bc3/rest_framework/generics.py#L97-L100) and 404 if it isn't found, or are you trying to circumvent the PermissionDenied exception?